src/cascadia/WindowsTerminal/WindowEmperor.cpp

```diff
@@ -18,600 +24,1015 @@ using namespace ::Microsoft::Console;
-//   - Setting up callbacks for when the number of windows changes.
-//   - Setting up the throttled func for layout persistence. Arguments:
-// - <none>
-void WindowEmperor::_becomeMonarch()
-{
-    // Add a callback to the window manager so that when the Monarch wants a new
-    // window made, they come to us
-    _manager.RequestNewWindow([this](auto&&, const Remoting::WindowRequestedArgs& args) {
-        _createNewWindowThread(args);
-    });
+    const auto parsedTarget = args.TargetWindow();
+    WindowingMode windowingBehavior = WindowingMode::UseNew;
+    uint64_t windowId = 0;
+    winrt::hstring windowName;
 
-    _createMessageWindow();
+    // Figure out the windowing behavior the caller wants
+    // and get the sanitized window ID (if any) and window name (if any).
+    if (parsedTarget.empty())
+    {
+        windowingBehavior = _app.Logic().Settings().GlobalSettings().WindowingBehavior();
+    }
+    else if (const auto opt = til::parse_signed<int64_t>(parsedTarget, 10))
+    {
+        // Negative window IDs map to WindowingMode::UseNew.
+        if (*opt > 0)
+        {
+            windowId = gsl::narrow_cast<uint64_t>(*opt);
+        }
+        else if (*opt == 0)
+        {
+            windowingBehavior = WindowingMode::UseExisting;
+        }
+    }
+    else if (parsedTarget == L"last")
+    {
+        windowingBehavior = WindowingMode::UseExisting;
+    }
+    // A window name of "new" maps to WindowingMode::UseNew.
+    else if (parsedTarget != L"new")
+    {
+        windowName = parsedTarget;
+    }
 
-    _setupGlobalHotkeys();
+    AppHost* window = nullptr;
 
-    // When the settings change, we'll want to update our global hotkeys and our
-    // notification icon based on the new settings.
-    _app.Logic().SettingsChanged([this](auto&&, const TerminalApp::SettingsLoadEventArgs& args) {
-        if (SUCCEEDED(args.Result()))
+    // Map from the windowing behavior, ID, and name to a window.
+    if (windowId)
+    {
+        window = GetWindowById(windowId);
+    }
+    else if (!windowName.empty())
+    {
+        window = GetWindowByName(windowName);
+    }
+    else
+    {
+        switch (windowingBehavior)
         {
-            _setupGlobalHotkeys();
-            _checkWindowsForNotificationIcon();
+        case WindowingMode::UseAnyExisting:
+            window = _mostRecentWindow();
+            break;
+        case WindowingMode::UseExisting:
+            _dispatchCommandlineCurrentDesktop(std::move(args));
+            return;
+        default:
+            break;
         }
-    });
+    }
 
-    // On startup, immediately check if we need to show the notification icon.
-    _checkWindowsForNotificationIcon();
+    if (window)
+    {
+        window->DispatchCommandline(std::move(args));
+    }
+    else
+    {
+        winrt::TerminalApp::WindowRequestedArgs request{ windowId, std::move(args) };
+        request.WindowName(std::move(windowName));

```

Comment:
Entirely sytlistically, it's a little weird that this isn't like `window = _mostRecentWindowCurrentDesktop(); break;`. It's the only early return in what is otherwise a very methodical function

Reply:
That's unfortunately necessary because `_mostRecentWindowCurrentDesktop` is a coroutine. We could of course turn this function into a coroutine as well and then do
```
window = co_await _mostRecentWindowCurrentDesktop();
break;
```
It's mostly a coincidence that I drew the boundary like that.